### PR TITLE
8263776: [JVMCI] add helper to perform Java upcalls

### PR DESCRIPTION
**Pull request by tkrodriguez**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263776](https://bugs.openjdk.java.net/browse/JDK-8263776): [JVMCI] add helper to perform Java upcalls

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/3067/head:pull/3067`
`$ git checkout pull/3067`

To update a local copy of the PR:
`$ git checkout pull/3067`
`$ git pull https://git.openjdk.java.net/jdk pull/3067/head`

### CODE REVIEW

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 1]
```diff
@@ -456,6 +456,107 @@ JRT_BLOCK_ENTRY(int, JVMCIRuntime::throw_class_cast_exception(JavaThread* thread
...
+      case T_SHORT:
+        return (jshort) value->i;
+      case T_INT:
+      case T_FLOAT:
+        return value->i;
+      case T_LONG:
+      case T_DOUBLE:
+        return value->j;
+      default:
+        ShouldNotReachHere();
```
Comment (vnkozlov) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:554]: I would prefer to see `fatal(msg)` which prints what is unexpected type.

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 2]
```diff
@@ -456,6 +456,107 @@ JRT_BLOCK_ENTRY(int, JVMCIRuntime::throw_class_cast_exception(JavaThread* thread
...
+
+  methodHandle mh(thread, method);
+  if (mh->size_of_parameters() > 1 && !mh->is_static()) {
+    THROW_MSG_0(vmSymbols::java_lang_IllegalArgumentException(), "Invoked method must be static and take at most one argument");
+  }
+
+  Symbol* signature = mh->signature();
+  JavaCallArguments jca(mh->size_of_parameters());
+  ArgumentPusher jap(signature, &jca, argument, mh->is_static());
+  JavaValue result(jap.get_ret_type());
```
Comment (vnkozlov) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:527]: `jap.get_ret_type()` is used several times in this code. Why not cache it in local variable?

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 3]
```diff
@@ -456,6 +456,107 @@ JRT_BLOCK_ENTRY(int, JVMCIRuntime::throw_class_cast_exception(JavaThread* thread
...
+  double next_double() {
+    guarantee(!_pushed, "one argument");
+    _pushed = true;
+    jvalue v;
+    v.j = _argument;
+    return v.d;
+  }
+
+ public:
+  ArgumentPusher(Symbol* signature, JavaCallArguments*  jca, jlong argument, bool is_static) : SignatureIterator(signature) {
```
Comment (vnkozlov) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:488]: `is_static` is not used.

Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp]: fixed

File: src/hotspot/share/jvmci/jvmciRuntime.cpp (hunk 2 above)
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp]: fixed

File: src/hotspot/share/jvmci/jvmciRuntime.cpp (hunk 1 above)
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp]: fixed

**Comment by tkrodriguez** [comment]
I had pushed the wrong version of this in the initial commit but it's now the correct one.  There were changes to SignatureIterator that needs accommodation and I'd pushed an older version of this fix.

**Review by vnkozlov — APPROVED** [review]
Good.

Please, run at least tier1 mach5 testing before push.

**Comment by dougxc** [comment]
Looks good to me.

**Comment by vnkozlov** [comment]
Thank you, for running tests.

**Pull request closed without merging**